lib/pbench/server/database/models/users.py

```diff
@@ -1,58 +1,121 @@
+
+
 class User(Database.Base):
     """User Model for storing user related details."""
 
     __tablename__ = "users"
 
     id = Column(Integer, primary_key=True, autoincrement=True)
-    username = Column(String(255), unique=True, nullable=False)
-    first_name = Column(String(255), unique=False, nullable=False)
-    last_name = Column(String(255), unique=False, nullable=False)
-    password = Column(LargeBinary(128), nullable=False)
-    registered_on = Column(DateTime, nullable=False, default=datetime.datetime.now())
-    email = Column(String(255), unique=True, nullable=False)
-    role = Column(Enum(Roles), unique=False, nullable=True)
-    auth_tokens = relationship("ActiveTokens", backref="users")
+    oidc_id = Column(String(255), unique=True, nullable=False)
+    username = Column(String(255), unique=True, nullable=True)
+    dataset_metadata = relationship(
+        "Metadata", back_populates="user", cascade="all, delete-orphan"
+    )
+    _roles = Column(String(255), unique=False, nullable=True)
+
+    @property
+    def roles(self):
+        if self._roles:
+            return [x for x in self._roles.split(";")]
+        else:
+            return []
+
+    @roles.setter
+    def roles(self, value):
+        if isinstance(value, list):
+            self._roles = ";".join(value)
+        else:
+            raise UserSqlError("Adding role", value, "Value is not a list")
 
     def __str__(self):
         return f"User, id: {self.id}, username: {self.username}"
 
     def get_json(self):
-        return {
-            "username": self.username,
-            "email": self.email,
-            "first_name": self.first_name,
-            "last_name": self.last_name,
-            "registered_on": self.registered_on,
-        }
+        return {"username": self.username, "id": self.id}
 
-    @staticmethod
-    def get_protected() -> list[str]:
-        """Return protected column names that are not allowed for external updates.
+    def _decode(self, exception: IntegrityError) -> Exception:
+        """Decode a SQLAlchemy IntegrityError to look for a recognizable UNIQUE
+        or NOT NULL constraint violation.
+
+        Return the original exception if it doesn't match.
+
+        Args:
+            exception : An IntegrityError to decode
 
-        `auth_tokens` is already protected from external updates via PUT api since
-        it is of type sqlalchemy relationship ORM package and not a sqlalchemy column.
+        Returns:
+            a more specific exception, or the original if decoding fails
         """
-        return ["registered_on", "id"]
+        # Postgres engine returns (code, message) but sqlite3 engine only
+        # returns (message); so always take the last element.
+        cause = exception.orig.args[-1]
+        if "UNIQUE constraint" in cause:
+            return UserDuplicate(self, cause)
+        elif "NOT NULL constraint" in cause:
+            return UserNullKey(self, cause)
+        return exception
```

Comment:
Oh, yuck:  is this the **_fourth_** cookie-cutter-copy of this code??!

I think we need to have some common SQL-related code which does things like decode `IntegrityError` and raises commonly-derived "Duplicate" and/or "NullKey" exception objects.  (We should be able to derive `UserError` from something other than `Exception` which encompasses the commonly-derived SQL exceptions.)

(If this is a reasonable thing to do, and you're willing to do it, @npalaska, then it should be done as a separate PR which is a precursor to this one.)

Reply:
Agree it needs to be in the separate PR but doesn't have to be a precursor, right? can be a follow-up PR to consolidate all the common SQL error handling from all the table definitions.

Reply:
If you put it in a precursor, then it only affects _three_ places in the code (and this PR gets the benefits of the refactoring for "free"); if you put it in a follow-on, then it has to change _**four**_ places in the code, and this PR gets no benefit.

Reply:
Maybe, but I don't think changing three places vs four places is much different in terms of the effort I need to put in. On the other hand, I also don't want to block this PR because of the refactoring of all the database table error handling.

Reply:
I'll fix this as a follow-up later.